### ISSUE TITLE
Private pages with group restrictions result in infinite loop between login url and target page

### ISSUE THREAD

**Issue by probabble** [wagtail/wagtail#4297]
### Issue Summary

We need a special setting to specify which template to render when the user does not have permission to view a page.

### Steps to Reproduce

Our site uses a custom login page that creates a session via an external web service. A new page configured in the admin to restrict access to certain groups redirects to that external web service. Since the user _is already logged in_ the service redirects them back to the page where they don't have permission, which redirects them back to the login page -- this loop continues until the maximum redirect threshold is hit.

The login page is probably not the best place to send an authenticated user who does not have the right permissions - how often will these users have multiple accounts with different permissions? It would be more appropriate to direct the user to a 401/403/404 page by default. We would benefit from the ability to specify a setting like `WAGTAIL_FRONTEND_UNAUTHORIZED_TEMPLATE`/`WAGTAIL_FRONTEND_UNAUTHORIZED_URL` to 
configure how the site responds when private page is requested without the right permissions (similar to how `WAGTAIL_FRONTEND_LOGIN_URL`/`WAGTAIL_FRONTEND_LOGIN_URL` work).

### Technical details

* Python version: 3.4.
* Django version: 1.11 
* Wagtail version: 1.13.1
* Browser version: all

**Comment by DanAtShenTech** [comment]
@probabble I also just discovered this issue when I needed to restrict access to a page to a specific group of users.  I think this issue should be classified as a bug.  Would you consider changing the name of this issue from 
> private pages with group restrictions don't work with custom WAGTAIL_FRONTEND_LOGIN_URL 

to
>Private pages with group restrictions result in infinite loop between login url and target page

or wording similar to that?  Obviously the first level of protection that comes with restricting pages to certain groups (requiring login) works, but the second level (checking group membership) is what causes the error because it redirects back to the login url when the group membership fails rather than to some url specified by the developer.

To me, using the words "don't work with custom WAGTAIL_FRONTEND_LOGIN_URL" in the title makes this issue seem less of a problem than it really is.  I literally CANNOT USE the group-level permissions feature as I had planned but, rather, have to do hard coding in the template in order to check for membership in a group and then render different HTML as appropriate.

Rather than having something in a settings file to specify where to redirect, perhaps a Wagtail `Page` property that can accept either a url path or a url name would be more appropriate since restrictions are placed on individual pages - something like `group_permission_failure_url = None`.  So, for example, if I specify a page that is to only be visible to members of the group 'Life Members', then I might set `group_permission_failure_url = "/must-be-life"` which would be a Wagtail page that is not visible in the navigation and which has the message "You must be a Life Member in order to view this page.".

[Link to documentation](https://docs.wagtail.io/en/v2.4/advanced_topics/privacy.html?highlight=private%20pages) included here for reference since it is currently silent on handling the situation of permissions failing on lack of group membership.

Hoping @gasman will comment and also determine whether this issue should be flagged as a bug.

**Comment by probabble** [comment]
I'm happy to amend the description to include your experience @DanAtShenTech.

**Comment by alxbridge** [comment]
I've also run into this issue: I think what's required is an extra check in _wagtail_hooks.py_'s `check_view_restrictions` function, for `request.user.is_authenticated`, and do something appropriate if so.

I'm currently working around this with an additional function that registers with the `before_serve_page` hook (as registered by `check_view_restrictions`) and raises a `PermissionDenied` if the user is already authenticated but doesn't belong to the necessary restricted group.

**Comment by matburnham** [comment]
Yeah, I've created a `wagtail_hooks.py` as suggested by alxbridge:

```
from wagtail.core import hooks
from wagtail.core.models import PageViewRestriction
from django.core.exceptions import PermissionDenied

@hooks.register('before_serve_page', order=-1)
def check_group(page, request, serve_args, serve_kwargs):
    if request.user.is_authenticated:
        for restriction in page.get_view_restrictions():
            if not restriction.accept_request(request):
                if restriction.restriction_type == PageViewRestriction.GROUPS:
                    current_user_groups = request.user.groups.all()
                    if not any(group in current_user_groups for group in restriction.groups.all()):
                        raise PermissionDenied
```

**Comment by ehmadzubair** [comment]
Really appreciate the work you're doing with Wagtail. Is this issue under somebody's radar?

**Comment by jeriox** [comment]
Also bumped into this issue and would be happy to see this addressed

**Comment by tbrlpld** [comment]
I just ran into this too. I find it very odd to redirect authenticated users to a login view. There seems to be some assumption that a user might have multiple accounts buried in this behavior. I wonder how many such users there really are. 

I think it would be more consistent do show raise "permission denied" for authenticated users (without the right group) and perform the redirect to the login view only for anonymous users.

**Comment by lionelwust** [comment]
Any fix planned for this one ?

**Comment by sonnybaker** [comment]
I'm keen to submit a PR for this, but am wondering if it might be better to allow users to specify a handler method? I'm thinking about a circumstance where it might be more helpful to serve a template or redirect for users explaining why they can't access certain content, falling back to `PermissionDenied` by default. Pseudo-code, but something like...

```
WAGTAIL_UNAUTHORISED_GROUP_HANDLER = 'foo.utils.unauthorised_group_handler'`
```

```
def unauthorised_group_handler(page, request, restriction):
    return redirect('foo:upgrade_your_membership')
```

```
if not request.user.is_authenticated:
    response = require_wagtail_login(next=request.get_full_path())
    add_never_cache_headers(response)
    return response
else:
    handler_path = getattr(
        settings,
        "WAGTAIL_UNAUTHORISED_GROUP_HANDLER",
        None,
    )
    if handler_path:
        handler = import_string(handler_path)
        return handler(page, request, restriction)
    else:
        raise PermissionDenied
       # or perhaps we set the default WAGTAIL_UNAUTHORISED_GROUP_HANDLER in Wagtail core to an
       # `unauthorised_group_handler` that just raises this and we drop the if/else?

```

**Comment by jogwen** [comment]
Just hit this issue myself. Many thanks to @alxbridge and @matburnham for the workaround. The workaround I used in the end and worked in Wagtail 5.2 is detailed at https://stackoverflow.com/questions/62731867/.
